jobs/quicksuggest-2-bq/quicksuggest_2_bq/main.py

```diff
@@ -0,0 +1,207 @@
+
+    # Retrieve the base_url for attachments
+    server_info = client.server_info()
+    attachments_base_url = server_info["capabilities"]["attachments"]["base_url"]
+
+    # Only consider "data" records, search for the following code in Merino
+    # for record in remote_settings_client.records_of_type("data".to_string())
+    data_records = [
+        record for record in client.get_records() if record["type"] == "data"
+    ]
+
+    # Make use of connection pooling because all requests go to the same host
+    requests_session = requests.Session()
+
+    suggestions = {}
+
+    for record in data_records:
+        attachment_url = f"{attachments_base_url}{record['attachment']['location']}"
+
+        response = requests_session.get(attachment_url)
+
+        if response.status_code != 200:
+            # Ignore unsuccessful requests for now
+            logging.error(
+                (
+                    "Failed to download attachment for record with ID '%s'."
+                    " Response status code %s."
+                ),
+                record["id"],
+                response.status_code,
+            )
+            continue
+
+        # Each attachment is a list of suggestion objects and each suggestion
+        # object contains a list of keywords. Load the suggestions into pydantic
+        # model instances to discard all fields which we don't care about here.
+        suggestions.update(
+            {
+                suggestion_data["id"]: KintoSuggestion(**suggestion_data)
+                for suggestion_data in response.json()
+            }
+        )
+
+    return suggestions
+
+
+def store_suggestions(
+    date: datetime.datetime,
+    destination_project: str,
+    destination_table_id: str,
+    kinto_suggestions: Dict[int, KintoSuggestion]
+):
+    """Get records, download attachments and return the suggestions."""
+
+    # Turn the suggestions into dicts and augment them with
+    # an insertion date.
+    suggestions = [
+        {**asdict(suggestion), "insert_date": date.date().isoformat()}
+        for suggestion in kinto_suggestions.values()
+    ]
+
+    client = bigquery.Client(project=destination_project)
+
+    job_config = bigquery.LoadJobConfig(
+        create_disposition="CREATE_IF_NEEDED",
+        schema=[
+            bigquery.SchemaField("insert_date", "DATE"),
+            bigquery.SchemaField("advertiser", "STRING"),
+            bigquery.SchemaField("click_url", "STRING"),
+            bigquery.SchemaField("iab_category", "STRING"),
+            bigquery.SchemaField("icon", "STRING"),
+            bigquery.SchemaField("impression_url", "STRING"),
+            bigquery.SchemaField("id", "INTEGER"),
+            bigquery.SchemaField("keywords", "STRING", "REPEATED"),
+            bigquery.SchemaField("title", "STRING"),
+            bigquery.SchemaField("url", "STRING"),
+        ],
+        source_format=bigquery.SourceFormat.NEWLINE_DELIMITED_JSON,
+        write_disposition="WRITE_APPEND",
```

Comment:
Our ETL best practices include making sure that jobs are idempotent, so we generally use WRITE_TRUNCATE since WRITE_APPEND could lead to duplicates if the job gets rerun. To support having a history, we should add `time_partitioning` here to make this a date-partitioned table. And for consistency with existing practice, I'd recommend naming the date field `submission_date`.

Reply:
I do generally think it's good to maintain a history table for ETL jobs like this, but since it's not possible (I assume) to fetch old versions of content from RemoteSettings, it may be fine to have this table simply represent the current state and not be partitioned. In which case we would not need a `--date` input parameter.

Reply:
This all makes sense, thanks for the detailed explanation! I was working under the assumption that there's value in storing the history in the table. While we cannot fetch it from RS itself, [it is available](https://github.com/mozilla-services/quicksuggest-rs/tree/main/source-data) (see the [relevant discussion](https://mozilla.slack.com/archives/G01NU8LUXPH/p1636991319275200?thread_ts=1636983055.268100&cid=G01NU8LUXPH)).

I'm ok with keeping this job simple, i.e. dropping the `--date`, and adding it back if we'll ever need it. Does that sound ok, @jklukas ?

Reply:
Yes, that sounds great.